utils/mgmt_system.py

```diff
@@ -71,6 +77,32 @@ def get_search_filter_spec(self, *args, **kwargs):
 
         return pfs
 
+    def __getattribute__(self, attr):
+        # fetch the attribute using parent class to avoid recursion
+        res = super(_PsphereClient, self).__getattribute__(attr)
+        # any callable (except login) is login protected
+        if callable(res) and attr not in ['login']:
+            # fetch the cached, login-protected version
+            try:
+                return self._cached_protect[res]
+            # or create it and save to cache
+            except KeyError:
+                # simple try-except wrapper that tries to log in on failure
+                def login_protect(client, o):
+                    def f(*args, **kwargs):
+                        try:
+                            return o(*args, **kwargs)
+                        except:
+                            logger.warning('{} api disconnected; logging back in'.format(
+                                client.server))
+                            client.login()
+                            return o(*args, **kwargs)
+                    return f
+                self._cached_protect[res] = login_protect(self, res)
+                return self._cached_protect[res]
```

Comment:
`login_protect` is a decorator. Assuming it is defined elsewhere, I think the cost of doing this would be pretty low (and even if not, it would be more readable by far):
```python
if callable(res) and attr != 'login':
    return login_protect(self, res)
```

Maybe rename `login_protect` to `login_retry_wrapper`, and the intent starts to become obvious.